otter/test/test_controller.py

```diff
@@ -1139,3 +1163,257 @@ def test_real_convergence_zero_delta(self):
+            (set_nova_metadata_item('server_id', *DRAINING_METADATA).intent,
+                lambda _: raise_(ValueError('oops!')))
+        ])
+        self.assertRaises(ValueError, self._remove, True, True, dispatcher)
+        self.assertEqual(dispatcher.sequence, [])
+
+    def test_checks_pass_replace_true_no_purge_success(self):
+        """
+        If all the checks pass, and purge is false, then autoscale-specific
+        metadata is removed from the server's metadata.  If this is successful,
+        then the whole effect is a success, and returns same state the function
+        was called with.
+        """
+        dispatcher = SequenceDispatcher([
+            (get_server_details('server_id').intent,
+                lambda _: self.server_details),
+            (EvictServerFromScalingGroup(scaling_group=self.group,
+                                         server_id='server_id'),
+                lambda _: None)
+        ])
+        result = self._remove(True, False, dispatcher)
+        self.assertEqual(result, self.state)
+        self.assertEqual(dispatcher.sequence, [])
+
+    def test_checks_pass_replace_false_no_purge_success(self):
+        """
+        If all the checks pass, and purge is true and replace is false, then
+        "DRAINING" is added to the server's metadata.  If this is successful,
+        then the whole effect is a success, and returns the state the function
+        was called with, with the desired value decremented.
+        """
+        old_desired = self.state.desired = 2
+        dispatcher = SequenceDispatcher([
+            (get_server_details('server_id').intent,
+                lambda _: self.server_details),
+            (GetScalingGroupInfo(tenant_id='tenant_id', group_id='group_id'),
+                lambda _: self.group_manifest_info),
+            (EvictServerFromScalingGroup(scaling_group=self.group,
+                                         server_id='server_id'),
+                lambda _: None)
+        ])
+        result = self._remove(False, False, dispatcher)
+        self.assertEqual(result, self.state)
+        self.assertEqual(result.desired, old_desired - 1)
+        self.assertEqual(dispatcher.sequence, [])
+
+    def test_checks_pass_replace_true_no_purge_failure(self):
+        """
+        If all the checks pass, and purge is true, then "DRAINING" is added
+        to the server's metadata.  If this fails, then the failure is
+        propagated and the state is not returned.
+        """
+        dispatcher = SequenceDispatcher([
+            (get_server_details('server_id').intent,
+                lambda _: self.server_details),
+            (EvictServerFromScalingGroup(scaling_group=self.group,
+                                         server_id='server_id'),
+                lambda _: raise_(ValueError('oops')))
+        ])
+        self.assertRaises(ValueError, self._remove, True, False, dispatcher)
+        self.assertEqual(dispatcher.sequence, [])
```

Comment:
I see you picked up on this :) I'm actually thinking of adding an explicit `SequenceDispatcher.completed` method that returns True when the dispatcher has been exhausted, since it turns out it's a really important part of using this (I've already had tests that continued to pass when I forgot/removed some code at the end of a function...)